### PR TITLE
Added a unit test that reproduces the expected count matrix for snRNA-Seq mode in CreateCountMatrix

### PR DESCRIPTION
**Pull request by kishorikonwar**
### Purpose
The purpose is to add a unit test that generates a set of alignments that creates the count matrix where the intronic alignments are counted for the snRNA-Seq mode.

- No issue is linked to this PR.

### Changes
Added a new test where a set of alignments are generated for a chromosome, precompute two expected count matrices if the alignments were run with single-cell and singe nuclei mode and then compare these with the data as processed by the sctools to show the equivalence of it.

- No changes.
The most important function is toe look at the new function 
test_count_matrix_with_introns in the src/sctools/test/test_count.py files and run the pytest

- No instructions.

### CODE REVIEW

File: src/sctools/count.py [hunk 1]
```diff
@@ -387,7 +388,13 @@ def from_mtx(
             newline delimited column index file
 
         Returns
-        -------
+        ----------------- Forwarded message ---------
```
Comment (barkasn) [inline src/sctools/count.py]: Please remove as appropriate

**Comment by kishorikonwar** [comment]
retest

File: src/sctools/gtf.py [hunk 2]
```diff
@@ -350,3 +350,56 @@ def extract_extended_gene_names(
         # Sort by starting location
         gene_locations[chromosome].sort(key=lambda x: x[0])
     return gene_locations
+
+
+def extract_gene_exons(
+    files: Union[str, List[str]] = "-", mode: str = "r", header_comment_char: str = "#"
+) -> Dict[str, List[tuple]]:
+    """Extract extended gene names from GTF file(s) and returns a map from gene names to the the
```
Comment (barkasn) [inline src/sctools/gtf.py]: 'the' repetition

File: src/sctools/test/test_barcode.py [hunk 3]
```diff
@@ -144,8 +144,7 @@ def tagged_bamfile():
         outbam,
     ]
     platform.TenXV2.attach_barcodes(args)
-    yield outbam
```
Comment (barkasn) [inline src/sctools/test/test_barcode.py]: Why are we changing this?

File: src/sctools/test/test_count.py [hunk 4]
```diff
@@ -813,3 +841,519 @@ def test_count_matrix_from_bam(
             )
         ]
     )
+
+
+def extract_gene_non_exons(
```
Comment (barkasn) [inline src/sctools/test/test_count.py]: Add heredoc description of this function

File: src/sctools/test/test_count.py [hunk 5]
```diff
@@ -813,3 +841,519 @@ def test_count_matrix_from_bam(
...
+
+    See Also
+    --------
+    count.from_sorted_tagged_bam
+    """
+
+    OUTPUT_PREFIX = "intronic_"
+    SYNTHETIC_SEQUENCE_LENGTH = 5
+    REFERENCE_SEQUENCE_NAME = "1"
+    # EXONIC_SEQUENCE_NAME = "EXONIC_SEQUENCE"
```
Comment (barkasn) [inline src/sctools/test/test_count.py]: Leftover comment?

**Comment by barkasn** [comment]
Did basic code review and discussed this with @kishori82 and its ok to go. I'd like some independent review of the code by an engineer please

File: src/sctools/gtf.py [hunk 6]
```diff
@@ -350,3 +350,56 @@ def extract_extended_gene_names(
...
+    mode : {'r', 'rb'}, optional
+        Open mode. If 'r', read strings. If 'rb', read bytes (default = 'r').
+    header_comment_char : str, optional
+        lines beginning with this character are skipped (default = '#')
+
+    Returns
+    -------
+    Dict[str, List[tuple]]
+        A dictionary of chromosome names mapping to a List of tuples, each containing
+        a the exons in the ascending order of the start positions.
```
Comment (gbggrant) [inline src/sctools/gtf.py]: 'the exons in ascending order of the start positions.'

File: src/sctools/test/test_barcode.py (hunk 3 above)
Comment (gbggrant) [reply src/sctools/test/test_barcode.py]: +1

File: src/sctools/test/test_count.py [hunk 7]
```diff
@@ -813,3 +841,519 @@ def test_count_matrix_from_bam(
...
+    row_index_output_filename = OUTPUT_PREFIX + "_row_index.npy"
+    col_index_output_filename = OUTPUT_PREFIX + "_col_index.npy"
+
+    def __init__(
+        self,
+        num_cells: int,
+        max_genes: int,
+        chromosomes_gene_exons: Dict[str, Dict[str, List[tuple]]],
+        chromosomes_gene_non_exons: Dict[str, List[tuple]],
+        rng_seed: int = 777,
```
Comment (gbggrant) [inline src/sctools/test/test_count.py]: Why provide a default here?  Can you pull it from time() or somesuch?

**Pull request merged**